napari/utils/events/containers/_set.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+from typing import Any, Iterable, Iterator, MutableSet, TypeVar
+
+from napari.utils.events import EmitterGroup
+
+_T = TypeVar("_T")
+
+
+class EventedSet(MutableSet[_T]):
+    """An unordered collection of unique elements.
+
+    Parameters
+    ----------
+    data : iterable, optional
+        Elements to initialize the set with.
+
+    Events
+    ------
+    added (value: _T)
+        emitted after an item is added to the set.  Will not be emitted if an
+        item was already in the set when added.
+    removed (value: _T)
+        emitted after an item is removed from the set.  Will not be emitted if
+        the item was not in the set when discarded.
```

Comment:
do we want to have different event for batch changes in particular `update()` ? Or even have `added` have a set of new values ? doing one event per insertion in particular in `update()` seem at risk of triggering accidental O(N^2) behavior; like if a gui element start to do N repainting on a set update.

Reply:
sure sounds good.  let's stick to one event name `added` ... do you think it should _always_ emit an iterable value, even it adds a single value, for the sake of simplicity on the receiving end?

Reply:
I think that yes, always an iterable should be good. That is what I would have done.

If you want you could even send a single change event with two added and remove attributes, which might make it easier to only hook one callback in general and avoid two events in `symmetric_difference_update`; I'm of two mind with this alternative though.

Reply:
updated to minimize events.  all events now have a `value` of type `set`